src/main/java/io/yupiik/maven/shade/transformer/BaseRelocatingTransformer.java

```diff
@@ -27,13 +27,8 @@
     protected String relocate(final String originalValue, final List<Relocator> relocators) {
         String newValue = originalValue;
         for (Relocator relocator : relocators) {
-            String value;
-            do {
-                value = newValue;
-                newValue = relocator.relocateClass(value);
-            }
-            while (!value.equals(newValue));
+            newValue = relocator.applyToSourceContent(newValue);
```

Comment:
this change looks unlikely since it can break existing usages, I guess we should inject the type of relocation to handle: CLASS (previous loop), SOURCE, PATH probably

wdyt?

Reply:
@Vlatombe issue with relocateClass is that it uses `replaceFirst` so the loop is needed as explained. `applyToSourceContent` uses a fully different algorithm which leads to a different output so we can't replace it like that if we want to still cover the original case, this is why I suggested to add a `replacementType` enum in the base transformer configuration (with probably CLASS as default for backward compatiblity). If it helps we can have a `ClassBaseRelocationTransformer` and `SourceContentBaseRelocationTransformer` which would pre-set this value. But high level, the feature is needed IMHO.

side note: also agree the replace usage is not the most efficient so happy to have relfection usage to extract `pattern` and `shadedPattern` to do it at once (pattern.replace) and drop the loop if you think the effort is worth it (it was okish on the resources we used it but I can envision that on big files it is not).

Reply:
dot handling is one trivial case which will not work the same in both cases, another one is the replacement of class AND path patterns for source call whereas for class it is only class.
One case I know which will work like that is for descriptors where class names are replaced with it and path names replaced in a downstream transformer (doing some more magic). It can be a rare case but I have to admit I'm not sure how often such a change will break so I'd prefer a not breaking change if possible, in particular since it does not prevent the feature and behavior you expect quite easily IMHO.

So overall either the enum config or a few more leaves (transformers) with different usages of the relocation can make it smooth for everyone (if so we should enhance the readme but that's it I guess).

Wdyt?

Reply:
I reworked the PR in two different transformers, but tbh I think the algorithms difference is difficult to explain if you don't know the implementation details.